static/js/typeahead_helper.js

```diff
@@ -24,25 +24,33 @@ export function get_cleaned_pm_recipients(query_string) {
 }
 
 export function build_highlight_regex(query) {
-    const regex = new RegExp("(" + _.escapeRegExp(query) + ")", "ig");
+    const regex = new RegExp("(" + _.escapeRegExp(query) + ")", "i");
     return regex;
 }
```

Comment:
It's not clear to me that this change is correct in all cases where we use `highlight_with_escaping_and_regex`, and would like to see that satisfactorily explained -- we did add the loop in the below logic for a reason, and I'm not sure what it was. 

Let's move this commit to a separate PR, perhaps with a #frontend thread to get more eyes on it. I can imagine one possibility is that we should be using passing an option to this function, and another (if we're matching individual words) is that the regular expression needs a negative lookbehind assertion about which types of characters sit before `query`.

Reply:
@timabbott The reason for the loop was probably for when the query has multiple words and so might have multiple highlighted parts (for example: **O**thello, the Moor **o**f Venice). I changed the logic to be much simpler and continue supporting multi word queries, as explained in a block comment. 

I also realised that for highlighting persons, we call the highlighting function per word of the name, while for streams, we do it for the entire stream name (which may be multi word). This resulted in the bug pointed out in the [above comment by Julia](https://github.com/zulip/zulip/pull/22307#pullrequestreview-1024711661), so I fixed that is a separate commit too! 

The PR is #22739